src/box/sql/build.c

```diff
@@ -2117,6 +2169,576 @@ sql_drop_table(struct Parse *parse_context)
 	sqlSrcListDelete(db, table_name_list);
 }
 
+/** Check is BOX constraint with given name exists. */
+static bool
+is_constraint_exists(struct tuple_constraint_def *cdef, uint32_t count,
+		     const char *name)
+{
+	for (uint32_t i = 0; i < count; ++i) {
+		if (strcmp(cdef[i].name, name) == 0)
+			return true;
+	}
+	return false;
+}
+
+/** Append a new BOX constraint to array of BOX constraints. */
+static struct tuple_constraint_def *
+constraint_append(struct Parse *parse, struct tuple_constraint_def *cdef,
+		  uint32_t count, struct tuple_constraint_def *fk)
+{
+	uint32_t size;
+	struct region *region = &parse->region;
+	struct tuple_constraint_def *new_cdef =
+		region_alloc_array(region, typeof(*new_cdef), count + 1, &size);
+	if (new_cdef == NULL) {
+		diag_set(OutOfMemory, size, "region_alloc_array", "new_cdef");
+		parse->is_aborted = true;
+		return NULL;
+	}
+	for (uint32_t i = 0; i < count; ++i)
+		new_cdef[i] = cdef[i];
+	new_cdef[count] = *fk;
+	return new_cdef;
+}
+
+/** Drop a BOX constraint from array of BOX constraints. */
+struct tuple_constraint_def *
+constraint_drop(struct Parse *parse, struct tuple_constraint_def *cdef,
+		uint32_t count, uint32_t cid)
+{
+	uint32_t size;
+	struct region *region = &parse->region;
+	struct tuple_constraint_def *new_cdef =
+		region_alloc_array(region, typeof(*new_cdef), count, &size);
+	if (new_cdef == NULL) {
+		diag_set(OutOfMemory, size, "region_alloc_array", "new_cdef");
+		parse->is_aborted = true;
+		return NULL;
+	}
+	for (uint32_t j = 0, k = 0; j < count + 1; ++j) {
+		if (j != cid)
+			new_cdef[k++] = cdef[j];
+	}
+	return new_cdef;
+}
+
+/** Return name of the new FOREIGN KEY constraint. */
+static const char *
+fk_name(struct Parse *parse)
```

Comment:
Lets not be too loose about static function names. I would suggest to be more specific and use `fk_unique_name_new()` or likewise.

Reply:
For now, I named it `sql_fk_name()`. The name is not checked for uniqueness, so I don't think it's worth mentioning.

Reply:
I dropped this change in current version.